### PR TITLE
Melt windows fix

### LINKED ISSUE #124: Windows Installation Issue
I am rather new to the Fortran environment in Windows and have encountered some trouble during installation. Even though I have followed the instructions provided on the Website for Windows (https://cambridge-iccs.github.io/FTorch/page/troubleshooting.html), the following error consistently shows up when running the "cmake --build ." command: 
 
1.	“fatal error LNK1181: cannot open input file ‘mkl_blacs_ilp64.lib’” (see attached screenshot) . Apparently I am not the only person with this issue (https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/Intel-oneAPI-Base-Toolkit-v-2024-1-Error-mkl-blacs-ilp64-lib/m-p/1587668#M3636), but also changing the environment variables didn’t do the job for me. For your understanding I have attached the full command lines that I have used for installation as txt to this E-Mail. I have removed this error with a very messy cleanup by copying and renaming that specific dll file in the intel compiler to test the further workflow, where I received more different errors (see 2.).
2.	When I try to run the Simplenet example the following errors show up: unresolved external symbol FTORCH_mp_TORCH_TENSOR_FROM_ARRAY_REAL32_1D referenced in function MAIN__ 
(and additional functions see second image attached). Is this error related to the first? Where is this variable defined?

The following changes to the installation have also been tried out but were not successful: 
1. running
cmake .. -G "NMake Makefiles" -DCMAKE_BUILD_TYPE=Release-DCMAKE_PREFIX_PATH="C:\Users\vbalmer\Anaconda3\envs\Py-F-02\Lib\site-packages\torch" -DCMAKE_INSTALL_PREFIX="C:\Users\vbalmer\AppData\Local\FTorch"
--> doesn't successfully run cmake command
2. including the old Fortran compiler ifort instead of ifx --> leads to same error as outlined above

Thank you very much for your help in advance!

Error_Screenshot (005)
[BuildInstructions_Clean.txt](https://github.com/Cambridge-ICCS/FTorch/files/15168045/BuildInstructions_Clean.txt)
Error_SimpleNet_Screenshot

### PR DESCRIPTION
**Pull request by TomMelt**
fixes #124 

Windows was not previously exporting the library symbols and creating the `dll` correctly.

This is now resolved by the following addition to `CMakeLists.txt`
https://github.com/Cambridge-ICCS/FTorch/blob/7856a33f9b81bad504da322c26f6f90c9b75a339/src/CMakeLists.txt#L6-L9

Furthemore, there was an issue that the library was built with a different integer size than that used when compiling example programs, despite no difference in compiler option. We have now _forced_ the library to use `int32` explicitly. This can be modified by the user, modifying the following line
https://github.com/Cambridge-ICCS/FTorch/blob/7856a33f9b81bad504da322c26f6f90c9b75a339/src/ftorch.fypp#L34

Tasks:
- [x] fix cmake windows build issue
- [x] fix `int64`/`int32` issue properly (we have decided to build int32 by default. Can be overridden by user.
- [x] update documentation to reflect new changes

### CODE REVIEW

**Comment by jatkinson1000** [comment]
Discussion:

- For the places we have int we replace by an explicit int64 (or similar).
- This will remove the issues on Windows where we get a mismatch.
- Ensure docs are updated to be clear users now need to specify the type of int.
- This will allow this to be closed, and relates to #124 and #140

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
This all looks reasonable and in agreement with what we discussed in the meeting, thanks!

The remaining thing to do is to document this, as mentioned in the checklist for the PR.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,11 @@ set(PROJECT_NAME FTorch)
 set(LIB_NAME ftorch)
 set(PACKAGE_VERSION 0.1)
 
+if (WIN32)
+    set (CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
+    set (BUILD_SHARED_LIBS TRUE)
+endif ()
+
```
Comment (jatkinson1000) [inline src/CMakeLists.txt:10]: A small comment to explain presence of this clause?

File: src/ftorch.fypp [hunk 2]
```diff
@@ -514,7 +516,7 @@ contains
     integer(c_int), parameter :: c_dtype = ${enum_from_prec(PREC)}$ !! Data type
     integer(c_int64_t)        :: strides(${RANK}$)                  !! Strides for accessing data
     integer(c_int), parameter :: ndims = ${RANK}$                   !! Number of dimension of input data
-    integer                   :: i
+    integer(ftorch_int)           :: i
```
Comment (jatkinson1000) [inline src/ftorch.fypp:519]:
```suggestion
    integer(ftorch_int)       :: i
```
Nit

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
Discussed in meeting, small comments on looking at the code, looks like just docs left.

**Comment by TomMelt** [comment]
> This all looks reasonable and in agreement with what we discussed in the meeting, thanks!
> 
> The remaining thing to do is to document this, as mentioned in the checklist for the PR.

thanks. Just done in https://github.com/Cambridge-ICCS/FTorch/pull/137/commits/d64358c7cb5d3a6c0c2427b1168a96d05cd01e94

File: src/CMakeLists.txt (hunk 1 above)
Comment (TomMelt) [reply src/CMakeLists.txt]: done in https://github.com/Cambridge-ICCS/FTorch/pull/137/commits/b3b50a16dc2563682470924e065e62e20f8bcad0

File: src/ftorch.fypp (hunk 2 above)
Comment (TomMelt) [reply src/ftorch.fypp]: I did one better. See https://github.com/Cambridge-ICCS/FTorch/pull/137/commits/594e9386338a4ba348b80fdbffe8ed3e07968ae7. Hope you like it :crossed_fingers:

File: pages/troubleshooting.md [hunk 3]
```diff
@@ -86,3 +86,31 @@ evaluation mode for efficiency.
...
+```fortran
+integer, parameter :: ftorch_int = int32 ! set integer size for FTorch library
+```
+
+We can use 64-bit integers by changing the above line to this
+```fortran
+integer, parameter :: ftorch_int = int64 ! set integer size for FTorch library
+```
+
+Finally, you will need to run `fypp` e.g.,
```
Comment (jwallwork23) [inline pages/troubleshooting.md:113]: I guess users may also need to install this because it isn't in the set of minimal requirements.

File: pages/troubleshooting.md [hunk 4]
```diff
@@ -86,3 +86,31 @@ evaluation mode for efficiency.
...
+```bash
+fypp src/ftorch.fypp > src/ftorch.F90
```
Comment (jwallwork23) [inline pages/troubleshooting.md:115]: Not important but FYI this also works without the `>`.

**Review by jwallwork23 — APPROVED** [review]
Thanks for adding the docs, @TomMelt. Just one minor request then good to go!

File: pages/troubleshooting.md (hunk 4 above)
Comment (TomMelt) [reply pages/troubleshooting.md:115]: Thanks @jwallwork23 TIL.

File: pages/troubleshooting.md (hunk 3 above)
Comment (TomMelt) [reply pages/troubleshooting.md]: resolved https://github.com/Cambridge-ICCS/FTorch/pull/137/commits/5cdfe22df277c84b410914ea4df4171c0cecc9df

File: pages/troubleshooting.md (hunk 4 above)
Comment (TomMelt) [reply pages/troubleshooting.md]: changed to your version... it saves 1 character :wink: https://github.com/Cambridge-ICCS/FTorch/pull/137/commits/5cdfe22df277c84b410914ea4df4171c0cecc9df

File: pages/troubleshooting.md [hunk 5]
```diff
@@ -86,3 +86,32 @@ evaluation mode for efficiency.
 These can be adjusted using the `requires_grad` and `is_training` optional arguments
 in the Fortran interface. See the [API procedures documentation](lists/procedures.html)
 for details.
+
+### How do I compile an int64 version of `ftorch` for large tensors?
+
+If your code uses large tensors (where large means more elements, in at least 1
+dimension, than the maximum size of a 32-bit integer (2,147,483,647)) you may
+need to compile `ftorch` with 64-bit integers. If you do not, you may receive a
+compile time error like the following:
```
Comment (jatkinson1000) [inline pages/troubleshooting.md:95]: Add small explanation as to why we made the decision.
```suggestion
Currently FTorch represents the number of elements in an array dimension using
32-bit integers. For most users this will be more than enough, but if your code
uses large tensors (where large means more than 2,147,483,647 elements
in any one dimension (the maximum value of a 32-bit integer)), you may you may
need to compile `ftorch` with 64-bit integers. If you do not, you may receive a
compile time error like the following:
```

**Review by jatkinson1000 — APPROVED** [review]
Couple of suggestions on phrasing, but not blocking.
Should be OK with this being added once #168 is approved and merged

**Comment by TomMelt** [comment]
the only test that fails is in the static analysis when running `fortitude`. This will be fixed separately by #205
